### PR TITLE
Fix script parsing to ignore closing tags in comments

### LINKED ISSUE #77: JSDOMParser errors on http://www.newyorker.com/magazine/2014/12/01/quiet-german
```
JSDOMParser error: expected '</lineargradient>'
JSDOMParser error: expected '</defs>'
JSDOMParser error: expected '</svg>'
JSDOMParser error: expected '</h1>'
JSDOMParser error: expected '</div>'
JSDOMParser error: expected '</div>'
JSDOMParser error: expected '</div>'
JSDOMParser error: expected '</div>'
JSDOMParser error: expected '</div>'
JSDOMParser error: expected '</body>'
JSDOMParser error: expected '</html>'
```

### PR DESCRIPTION
**Pull request by gijsk**
r? @n1k0 and/or @leibovic

This should fix #77, but I've not tested that, just the reduced testcase in this commit.

### CODE REVIEW

File: JSDOMParser.js [hunk 1]
```diff
@@ -919,14 +919,59 @@
...
 
     readScript: function (node) {
-      var index = this.html.indexOf("</script>", this.currentChar);
-      if (index === -1) {
-        index = this.html.length;
+      while (this.currentChar < this.html.length) {
+        var c = this.nextChar();
+        var nextC = this.peekNext();
+        if (c === "<") {
+          if (nextC === "!" || nextC === "?") {
```
Comment (n1k0) [inline JSDOMParser.js]: I may be missing something obvious, but what type of comments `?` does match? :)

Comment (n1k0) [inline JSDOMParser.js]: Bah, just learned about the `<?…?>` html comment type. Never too late I guess.

**Comment by n1k0** [comment]
Note: I'd expect tests to be added to the JSDOMParser suite instead of only relying on Readability ones.

**Comment by gijsk** [comment]
> Note: I'd expect tests to be added to the JSDOMParser suite instead of only relying on Readability ones.

Good point, let me write some and get back here.

File: JSDOMParser.js [hunk 2]
```diff
@@ -919,14 +919,59 @@
...
-      var txt = new Text();
-      txt.textContent = this.html.substring(this.currentChar, index === -1 ? this.html.length : index);
-      node.appendChild(txt);
-      this.currentChar = index;
+    },
+
+    discardNextComment: function() {
+      if (this.match("--")) {
+        this.discardTo("-->");
+      } else {
```
Comment (n1k0) [inline JSDOMParser.js]: Shouldn't we discardTo(`?>`) as well if `?` is found? This test case fails:

```js
  it("should strip comments within script tags", function() {
    var html = '<script><?Silly test > <script src="foo.js"></script>?></script>';
    var doc = new JSDOMParser().parse(html);
    expect(doc.firstChild.tagName).eql("SCRIPT");
    expect(doc.firstChild.textContent).eql("");
  });
```

Comment (gijsk) [inline JSDOMParser.js]: Probably! This is copied from the existing code, actually... so I guess that was also broken before. Writing tests atm, that'll help here.

Comment (gijsk) [inline JSDOMParser.js]: Err, actually the parser spec says: http://www.w3.org/TR/html5/syntax.html#bogus-comment-state

> Consume every character up to and including the first ">" (U+003E) character or the end of the file (EOF), whichever comes first. Emit a comment token whose data is the concatenation of all the characters starting from and including the character that caused the state machine to switch into the bogus comment state, up to and including the character immediately before the last consumed character (i.e. up to the character just before the U+003E or EOF character)

And this is triggered here: http://www.w3.org/TR/html5/syntax.html#tag-open-state

Comment (gijsk) [inline JSDOMParser.js]: In other words, AFAICT the current behaviour is correct...

**Comment by n1k0** [comment]
LGTM especially with the tests added and a green build :)

**Comment by gijsk** [comment]
r+ from @n1k0 over IRC, so merging this in. :)

**Pull request merged**